File: bindings/lua/tests/language_bindings_test.lua
-- Trigger linter refresh
local luaunit = require('luaunit')
local satox = require('satox')
local blockchain = require('satox.blockchain')
local transactions = require('satox.transactions')
local assets = require('satox.assets')
local wallet = require('satox.wallet')
local quantum = require('satox.quantum')
local ipfs = require('satox.ipfs')
local security = require('satox.security')

TestLanguageBindings = {}

function TestLanguageBindings:setUp()
    -- Initialize all managers
    self.manager = satox.SatoxManager.new()
    self.blockchainManager = blockchain.BlockchainManager.new()
    self.transactionManager = transactions.TransactionManager.new()
    self.assetManager = assets.AssetManager.new()
    self.walletManager = wallet.WalletManager.new()
    self.quantumManager = quantum.QuantumManager.new()
    self.ipfsManager = ipfs.IPFSManager.new()
    self.securityManager = security.SecurityManager.new()
end

function TestLanguageBindings:tearDown()
    -- Cleanup
    if self.manager then
        self.manager:shutdown()
    end
end

-- Basic Functionality Tests
function TestLanguageBindings:testInitialization()
    -- Test SatoxManager initialization
    luaunit.assertFalse(self.manager.initialized)
    self.manager:initialize()
    luaunit.assertTrue(self.manager.initialized)
    
    -- Test double initialization
    luaunit.assertError(function()
        self.manager:initialize()
    end, "SatoxManager already initialized")
end

-- API Manager Tests
function TestLanguageBindings:testAPIManager()
    -- Register API Manager component
    local apiManager = {
        initialize = function() return true end,
        shutdown = function() return true end
    }
    self.manager:registerComponent("api", apiManager)
    
    -- Test component registration
    local component = self.manager:getComponent("api")
    luaunit.assertNotNil(component)
    
    -- Test component unregistration
    self.manager:unregisterComponent("api")
    luaunit.assertError(function()
        self.manager:getComponent("api")
    end)
end

-- Wallet Manager Tests
function TestLanguageBindings:testWalletManager()
    -- Initialize required managers
    self.blockchainManager:initialize()
    self.quantumManager:initialize()
    self.walletManager:initialize(self.blockchainManager, self.quantumManager)
    
    -- Test wallet creation
    local wallet = self.walletManager:createWallet("test-wallet", "Test wallet")
    luaunit.assertNotNil(wallet)
    luaunit.assertEquals(wallet.name, "test-wallet")
    
    -- Test wallet retrieval
    local retrievedWallet = self.walletManager:getWallet(wallet.id)
    luaunit.assertNotNil(retrievedWallet)
    luaunit.assertEquals(retrievedWallet.id, wallet.id)
    
    -- Test wallet update
    local updatedWallet = self.walletManager:updateWallet(wallet.id, {
        description = "Updated description"
    })
    luaunit.assertEquals(updatedWallet.description, "Updated description")
end

-- IPFS Manager Tests
function TestLanguageBindings:testIPFSManager()
    -- Initialize IPFS Manager
    self.ipfsManager:initialize("127.0.0.1", 5001)
    
    -- Test data storage
    local data = "test data"
    local hash, err = self.ipfsManager:addData(data)
    luaunit.assertNotNil(hash)
    luaunit.assertNil(err)
    
    -- Test data retrieval
    local retrievedData, err = self.ipfsManager:getData(hash)
    luaunit.assertNotNil(retrievedData)
    luaunit.assertNil(err)
    
    -- Test pinning
    local success, err = self.ipfsManager:pinHash(hash)
    luaunit.assertTrue(success)
    luaunit.assertNil(err)
end

-- Quantum Manager Tests
function TestLanguageBindings:testQuantumManager()
    -- Initialize Quantum Manager
    self.quantumManager:initialize()
    
    -- Test key pair generation
    local keyPair = self.quantumManager:generate_key_pair()
    luaunit.assertNotNil(keyPair)
    luaunit.assertNotNil(keyPair.public_key)
    luaunit.assertNotNil(keyPair.private_key)
    
    -- Test encryption/decryption
    local data = "test data"
    local encrypted = self.quantumManager:encrypt(keyPair.public_key, data)
    luaunit.assertNotNil(encrypted)
    
    local decrypted = self.quantumManager:decrypt(keyPair.private_key, encrypted)
    luaunit.assertEquals(decrypted, data)
    
    -- Test signing/verification
    local signature = self.quantumManager:sign(keyPair.private_key, data)
    luaunit.assertNotNil(signature)
    
    local verified = self.quantumManager:verify(keyPair.public_key, data, signature)
    luaunit.assertTrue(verified)
end

-- Post-Quantum Algorithms Tests
function TestLanguageBindings:testPostQuantumAlgorithms()
    -- Test NTRU algorithm
    local keyPair = self.quantumManager:generate_key_pair("NTRU")
    luaunit.assertNotNil(keyPair)
    
    -- Test BIKE algorithm
    local bikeKeyPair = self.quantumManager:generate_key_pair("BIKE")
    luaunit.assertNotNil(bikeKeyPair)
    
    -- Test HQC algorithm
    local hqcKeyPair = self.quantumManager:generate_key_pair("HQC")
    luaunit.assertNotNil(hqcKeyPair)
end

-- Transaction Manager Tests
function TestLanguageBindings:testTransactionManager()
    -- Initialize required managers
    self.blockchainManager:initialize()
    self.quantumManager:initialize()
    self.transactionManager:initialize(self.blockchainManager, self.quantumManager)
    
    -- Test transaction creation
    local inputs = {
        {address = "addr1", amount = 100}
    }
    local outputs = {
        {address = "addr2", amount = 90}
    }
    
    local transaction = self.transactionManager:createTransaction(inputs, outputs)
    luaunit.assertNotNil(transaction)
    luaunit.assertEquals(#transaction.inputs, 1)
    luaunit.assertEquals(#transaction.outputs, 1)
    
    -- Test transaction signing
    local keyPair = self.quantumManager:generate_key_pair()
    local signedTransaction = self.transactionManager:signTransaction(transaction, keyPair.private_key)
    luaunit.assertNotNil(signedTransaction.signature)
end

-- Error Handling Tests
function TestLanguageBindings:testErrorHandling()
    -- Test invalid initialization
    luaunit.assertError(function()
        self.walletManager:initialize(nil, nil)
    end, "BlockchainManager is required")
    
    -- Test invalid wallet creation
    luaunit.assertError(function()
        self.walletManager:createWallet(nil)
    end, "Invalid wallet name")
    
    -- Test invalid transaction creation
    luaunit.assertError(function()
        self.transactionManager:createTransaction(nil, nil)
    end, "Inputs and outputs are required")
end

-- Concurrency Tests
function TestLanguageBindings:testConcurrency()
    -- Initialize managers
    self.manager:initialize()
    self.blockchainManager:initialize()
    self.quantumManager:initialize()
    self.walletManager:initialize(self.blockchainManager, self.quantumManager)
    
    -- Create multiple wallets concurrently
    local wallets = {}
    for i = 1, 10 do
        local wallet = self.walletManager:createWallet("wallet" .. i)
        table.insert(wallets, wallet)
    end
    
    -- Verify all wallets were created
    luaunit.assertEquals(#wallets, 10)
    for _, wallet in ipairs(wallets) do
        luaunit.assertNotNil(wallet)
        luaunit.assertNotNil(wallet.id)
    end
end

-- Performance Tests
function TestLanguageBindings:testPerformance()
    -- Initialize managers
    self.manager:initialize()
    self.blockchainManager:initialize()
    self.quantumManager:initialize()
    self.walletManager:initialize(self.blockchainManager, self.quantumManager)
    
    -- Test wallet creation performance
    local startTime = os.clock()
    for i = 1, 100 do
        self.walletManager:createWallet("perf_wallet" .. i)
    end
    local endTime = os.clock()
    local duration = endTime - startTime
    
    -- Expect to create at least 10 wallets per second
    luaunit.assertLessThan(duration, 10)
end

-- Recovery Tests
function TestLanguageBindings:testRecovery()
    -- Initialize managers
    self.manager:initialize()
    self.blockchainManager:initialize()
    self.quantumManager:initialize()
    self.walletManager:initialize(self.blockchainManager, self.quantumManager)
    
    -- Create a wallet
    local wallet = self.walletManager:createWallet("recovery_wallet")
    luaunit.assertNotNil(wallet)
    
    -- Simulate shutdown
    self.manager:shutdown()
    
    -- Reinitialize and verify recovery
    self.manager:initialize()
    self.blockchainManager:initialize()
    self.quantumManager:initialize()
    self.walletManager:initialize(self.blockchainManager, self.quantumManager)
    
    local recoveredWallet = self.walletManager:getWallet(wallet.id)
    luaunit.assertNotNil(recoveredWallet)
    luaunit.assertEquals(recoveredWallet.id, wallet.id)
end

-- Run tests
os.exit(luaunit.LuaUnit.run()) 

File: bindings/javascript/tests/language_bindings_test.js
const { expect } = require('chai');
const QuantumManager = require('../src/quantum/quantum_manager');
const PostQuantumAlgorithms = require('../src/quantum/post_quantum_algorithms');
const AssetManager = require('../src/assets/AssetManager');
const KeyStorage = require('../src/quantum/key_storage');
const HybridEncryption = require('../src/quantum/hybrid_encryption');

describe('Language Bindings', () => {
    let quantumManager;
    let postQuantumAlgorithms;
    let assetManager;
    let keyStorage;
    let hybridEncryption;

    beforeEach(() => {
        quantumManager = new QuantumManager();
        postQuantumAlgorithms = new PostQuantumAlgorithms();
        assetManager = new AssetManager();
        keyStorage = new KeyStorage();
        hybridEncryption = new HybridEncryption();
    });

    afterEach(() => {
        if (quantumManager) quantumManager.destroy();
        if (postQuantumAlgorithms) postQuantumAlgorithms.destroy();
        if (keyStorage) keyStorage.destroy();
        if (hybridEncryption) hybridEncryption.destroy();
    });

    // Basic Functionality Tests
    describe('Basic Functionality', () => {
        it('should initialize all managers', () => {
            expect(() => quantumManager.initialize()).to.not.throw();
            expect(() => postQuantumAlgorithms.isAlgorithmAvailable('NTRU')).to.not.throw();
            expect(() => assetManager.initialize()).to.not.throw();
            expect(() => keyStorage.initialize()).to.not.throw();
            expect(() => hybridEncryption.initialize()).to.not.throw();
        });

        it('should handle double initialization', () => {
            quantumManager.initialize();
            expect(() => quantumManager.initialize()).to.throw('Already initialized');
        });
    });

    // Quantum Manager Tests
    describe('Quantum Manager', () => {
        beforeEach(() => {
            quantumManager.initialize();
        });

        it('should generate key pairs', () => {
            const keyPair = quantumManager.generateKeyPair();
            expect(keyPair).to.have.property('publicKey');
            expect(keyPair).to.have.property('privateKey');
            expect(keyPair.publicKey).to.be.instanceof(Buffer);
            expect(keyPair.privateKey).to.be.instanceof(Buffer);
        });

        it('should encrypt and decrypt data', () => {
            const keyPair = quantumManager.generateKeyPair();
            const data = Buffer.from('test data');
            
            const encrypted = quantumManager.encrypt(keyPair.publicKey, data);
            expect(encrypted).to.be.instanceof(Buffer);
            
            const decrypted = quantumManager.decrypt(keyPair.privateKey, encrypted);
            expect(decrypted).to.be.instanceof(Buffer);
            expect(decrypted.toString()).to.equal('test data');
        });

        it('should sign and verify data', () => {
            const keyPair = quantumManager.generateKeyPair();
            const data = Buffer.from('test data');
            
            const signature = quantumManager.sign(keyPair.privateKey, data);
            expect(signature).to.be.instanceof(Buffer);
            
            const verified = quantumManager.verify(keyPair.publicKey, data, signature);
            expect(verified).to.be.true;
        });
    });

    // Post-Quantum Algorithms Tests
    describe('Post-Quantum Algorithms', () => {
        it('should check algorithm availability', () => {
            expect(postQuantumAlgorithms.isAlgorithmAvailable('NTRU')).to.be.true;
            expect(postQuantumAlgorithms.isAlgorithmAvailable('BIKE')).to.be.true;
            expect(postQuantumAlgorithms.isAlgorithmAvailable('HQC')).to.be.true;
        });

        it('should generate key pairs for different algorithms', () => {
            const algorithms = ['NTRU', 'BIKE', 'HQC'];
            algorithms.forEach(algorithm => {
                const keyPair = postQuantumAlgorithms.generateKeyPair(algorithm);
                expect(keyPair).to.have.property('publicKey');
                expect(keyPair).to.have.property('privateKey');
            });
        });

        it('should sign and verify with different algorithms', () => {
            const algorithms = ['NTRU', 'BIKE', 'HQC'];
            algorithms.forEach(algorithm => {
                const keyPair = postQuantumAlgorithms.generateKeyPair(algorithm);
                const data = Buffer.from('test data');
                
                const signature = postQuantumAlgorithms.sign(algorithm, keyPair.privateKey, data);
                expect(signature).to.be.instanceof(Buffer);
                
                const verified = postQuantumAlgorithms.verify(algorithm, keyPair.publicKey, data, signature);
                expect(verified).to.be.true;
            });
        });
    });

    // Asset Manager Tests
    describe('Asset Manager', () => {
        beforeEach(() => {
            assetManager.initialize();
        });

        it('should create and validate assets', () => {
            const asset = {
                id: 'test-asset',
                name: 'Test Asset',
                totalSupply: 1000
            };
            
            expect(assetManager.validateAsset(asset)).to.be.true;
        });

        it('should transfer assets', () => {
            const assetId = 'test-asset';
            const from = 'address1';
            const to = 'address2';
            const amount = 100;

            assetManager.assets.set(assetId, {
                id: assetId,
                name: 'Test Asset',
                totalSupply: 1000
            });
            assetManager.balances.set(`${assetId}:${from}`, 200);

            const transfer = assetManager.transferAsset(assetId, from, to, amount);
            expect(transfer).to.have.property('assetId', assetId);
            expect(transfer).to.have.property('from', from);
            expect(transfer).to.have.property('to', to);
            expect(transfer).to.have.property('amount', amount);
        });

        it('should get asset history', () => {
            const assetId = 'test-asset';
            assetManager.assets.set(assetId, {
                id: assetId,
                name: 'Test Asset',
                totalSupply: 1000
            });
            assetManager.balances.set(`${assetId}:address1`, 100);

            const history = assetManager.getAssetHistory(assetId);
            expect(history).to.have.property('asset');
            expect(history).to.have.property('transfers');
            expect(history.transfers).to.be.an('array');
        });
    });

    // Error Handling Tests
    describe('Error Handling', () => {
        it('should handle invalid initialization', () => {
            expect(() => {
                const invalidManager = new QuantumManager();
                invalidManager.manager = null;
                invalidManager.initialize();
            }).to.throw('Failed to initialize quantum manager');
        });

        it('should handle invalid key pairs', () => {
            expect(() => {
                quantumManager.encrypt(Buffer.from('invalid'), Buffer.from('data'));
            }).to.throw('Failed to encrypt data');
        });

        it('should handle invalid assets', () => {
            expect(() => {
                assetManager.validateAsset(null);
            }).to.throw('Invalid asset');
        });
    });

    // Concurrency Tests
    describe('Concurrency', () => {
        it('should handle concurrent key generation', async () => {
            const promises = Array(10).fill().map(() => 
                quantumManager.generateKeyPair()
            );
            
            const keyPairs = await Promise.all(promises);
            expect(keyPairs).to.have.length(10);
            keyPairs.forEach(keyPair => {
                expect(keyPair).to.have.property('publicKey');
                expect(keyPair).to.have.property('privateKey');
            });
        });

        it('should handle concurrent asset transfers', async () => {
            const assetId = 'test-asset';
            assetManager.assets.set(assetId, {
                id: assetId,
                name: 'Test Asset',
                totalSupply: 1000
            });
            assetManager.balances.set(`${assetId}:address1`, 500);

            const promises = Array(5).fill().map((_, i) => 
                assetManager.transferAsset(assetId, 'address1', `address${i + 2}`, 50)
            );
            
            const transfers = await Promise.all(promises);
            expect(transfers).to.have.length(5);
            transfers.forEach(transfer => {
                expect(transfer).to.have.property('amount', 50);
            });
        });
    });

    // Performance Tests
    describe('Performance', () => {
        it('should generate key pairs efficiently', () => {
            const startTime = Date.now();
            for (let i = 0; i < 100; i++) {
                quantumManager.generateKeyPair();
            }
            const endTime = Date.now();
            const duration = endTime - startTime;
            
            // Expect to generate at least 10 key pairs per second
            expect(duration).to.be.lessThan(10000);
        });

        it('should handle asset transfers efficiently', () => {
            const assetId = 'test-asset';
            assetManager.assets.set(assetId, {
                id: assetId,
                name: 'Test Asset',
                totalSupply: 1000
            });
            assetManager.balances.set(`${assetId}:address1`, 1000);

            const startTime = Date.now();
            for (let i = 0; i < 100; i++) {
                assetManager.transferAsset(assetId, 'address1', `address${i + 2}`, 1);
            }
            const endTime = Date.now();
            const duration = endTime - startTime;
            
            // Expect to process at least 10 transfers per second
            expect(duration).to.be.lessThan(10000);
        });
    });

    // Recovery Tests
    describe('Recovery', () => {
        it('should recover from shutdown', () => {
            quantumManager.initialize();
            const keyPair = quantumManager.generateKeyPair();
            quantumManager.shutdown();
            
            quantumManager.initialize();
            const newKeyPair = quantumManager.generateKeyPair();
            expect(newKeyPair).to.have.property('publicKey');
            expect(newKeyPair).to.have.property('privateKey');
        });

        it('should recover from errors', () => {
            assetManager.initialize();
            const assetId = 'test-asset';
            
            try {
                assetManager.transferAsset(assetId, 'address1', 'address2', 100);
            } catch (error) {
                expect(error.message).to.equal('Asset not found');
            }
            
            assetManager.assets.set(assetId, {
                id: assetId,
                name: 'Test Asset',
                totalSupply: 1000
            });
            assetManager.balances.set(`${assetId}:address1`, 200);
            
            const transfer = assetManager.transferAsset(assetId, 'address1', 'address2', 100);
            expect(transfer).to.have.property('amount', 100);
        });
    });
}); 